vulcanai/models/metrics.py

```diff
@@ -109,7 +109,7 @@ def extract_class_labels(in_matrix):
         # For one-hot encoded entries
         if in_matrix.shape[1] > 1:
             return np.argmax(in_matrix, axis=1)
-        # For binary entries
+        # For single value entries
         elif in_matrix.shape[1] == 1:
```

Comment:
Should we still keep this? perhaps we don't want to round without the user knowing? this was normally for single valued probabilities but continuous values would fall under this category

Reply:
Hmm, well I proposed a solution in my latest commit. You can now provide a callable to forward_pass. I don't think I need to expand it to get_score and run_test, as we'll always just score on non-transformed outputs? Or is that not true and there's a use case fro transforming and then scoring in the single continuous case (I know we do for one hot obvs)?

Reply:
ooooh I really like that transform_callable solution. I do think we should allow the score to be dependent on the transformed outputs but then the transform function would need to be done with Pytorch methods so backdrop can work well. This seems like it would lend itself well to the custom loss functionality no @nirtiac ?

Reply:
@rfratila yes it does. do we want that in this pull request as well?

Reply:
it might be a little outside the scope of this PR and a little more experimental so if you think we could separate this to another PR than I think that would be better no?

Reply:
I agree, @rfratila. In that case then I believe that this PR is finally ready.

Reply:
Yes. I would like @joemehltretter to do one last check to see if he is still running into eh bug he was before to make sure it was resolved. Then we are good to go! Thanks

Reply:
@rfratila I have come to the conclusion that we don't need to write any additional code in order to allow what we've been describing above. All the user would need to do is define a custom pred_activation that calls the same transform_callable function as well as a custom loss function (which well accept outputs passed through pred_activation). I don't think we would use this often enough to justify writing further wrappers, nor do I know what that would be. I could work on an example, however.

Reply:
wait so are you removing the transform_callable?

Reply:
No, that's still in there in what was convert_to_class and is now transform_outputs (other suggestions accepted)

Reply:
Ah ok and you would be able to pass a modified Softmax transform there to get the class outputs?

Reply:
yes

Reply:
Ok great. Did we want to have that specific case already implemented so it’s easy to use or nah? @joemehltretter I’d like to hear your thoughts as well on this generalized transform method.

Reply:
@rfratila @nirtiac So, `convert_to_class` parameter for `forward_pass` is not `transform_outputs`?

Reply:
It's not transform_outputs. But if you don't provide a transform_callable then it converts to class? Is that too opaque?

Reply:
What is an example of `transform_callable`? Just so I can be clear in my understanding

Reply:
@nirtiac I think that might be too much magic to do it automatically. Perhaps there’s something we can do to activate it easily but still manual?

Reply:
What's "it" in this case @rfratila? @joemehltretter a very simple example is passing np.round. Take a look at the end of predit_single_value_example.py

Reply:
By “it” I’m referring to the class conversion transform by default if nothing is provided

Reply:
well I can put that back to being a parameter

Reply:
The case you gave makes sense, I think it just depends on documentation of it because by itself it may not be easily understood. I do have a `forward_pass` case up and running and ready to test. I set it up last night so can test this tonight.

Reply:
The only thing I have to say is we should update the documentation with examples so it's clear for forward_pass so the new parameters are clearly understood. Other than that it's good!

Reply:
@nirtiac would you be able to take care of this ^ thanks :)

Reply:
dooone @rfratila

Reply:
Would you be able to update the run_test functions? They currently only have:
‘’’
raw_predictions = network.forward_pass(
            data_loader=data_loader,
            convert_to_class=False)
‘’’

Reply:
ahhhh good catch. they didn't fail cause the basic params are still there, you can provide extras at no consequence.

Reply:
Ah ok I thought we removed convert_to_class

Reply:
yeah we did, but we have kwargs, so it just ate it up

Reply:
Ok done, note that I don't propagate any transform_callable to run_test in the multi case cause we need that to be class converted one-hot output values. I did propagate it in the single value case. I did document that, so I think it's good to go. I personally re-tested everything and it all seems reasonable, but I guess it would be good if @joemehltretter used the new and improved run_test.

Reply:
Have you updated the cross validate as well? That calls run test

Reply:
ack, yes, thanks, done.

Reply:
I did a repo wide search and we have some calls in test_cnn.py, test_dnn.py, test_ensemble.py that would need updating as well.

Reply:
@joemehltretter would you be able to test it now with the new run_test and circle back around to us with any feedback? thanks :)

Reply:
@nirtiac @rfratila For `run_test` in `basenetwork` it doesn't match the parameters for `run_test` in metrics, why is that? Because when I call `run_test` after I train my model it doesn't have the transform parameters

Reply:
ack, you're right. I keep getting overenthusiastic thinking it's done. I'll fix that and do a thorough sweep of all usages everywhere.

Reply:
Ok, now cross_validate and run_test in basenetwork match in terms of parameters, and my search for all usages of run_test and forward_pass contain the new parameters.

Reply:
Thank you @nirtiac :) @joemehltretter can you confirm that your tests work well with these new updates?

Reply:
Just tested, all good.

Reply:
Documentation looks good as well

Reply:
Awesome thanks! @nirtiac are there any final cleanups you need to do?

Reply:
I pushed some changes so it looks good to me. If you don't have anything else to do I think it's ready to merge! If you can make sure the last commit is properly tagged so when merged to master (or the master merge commit is tagged) so the pypi build is good, that would be fantastic!

Reply:
Nope, I'm done. Can you approve so I can merge @rfratila?